### ISSUE TITLE
[clang] user-defined conversions failed.

### ISSUE THREAD

**Issue by changkhothuychung** [llvm/llvm-project#132600]
Given this program 

```cpp
struct Alloc{}; 

struct MyString
{
    MyString();
    MyString(MyString&&, const Alloc&); 
    MyString(MyString&&); 
    MyString(const MyString&, const Alloc&); 
    MyString(const MyString&); 
};

struct A
{
    operator const MyString() const; 
};

int main() {
    MyString s(A{}, Alloc{}); 
    return 0;
}
```

Clang reports error `no viable conversion from 'A' to 'MyString'`, but gcc doesn't complain about it. The issue goes away when I change `operator const` to `operator`. Is this a bug in clang? If yes, I would like to work on this issue. 

link to compiler explorer: https://godbolt.org/z/xKfs8GoPr

**Comment by changkhothuychung** [comment]
was using lldb to go through the program, since we have const `operator const`, so when we perform the user defined conversion from A, shouldn't `ToType` have const qualification in this line of code? https://github.com/llvm/llvm-project/blob/main/clang/lib/Sema/SemaOverload.cpp#L1665

However, when I try to inspect `isConstQualified` of ToType, it returns false. 

The overload `MyString(MyString&&, const Alloc&)` is selected in the viable list and finally selected as the best overload, but I think it shouldnt be in the viable list, as the conversion from `const MyString` to `MyString&&` is ill-formed.

**Comment by shafik** [comment]
Simpler reproducer:  https://godbolt.org/z/nqMd4ExPv

```cpp
struct Alloc{}; 

struct MyString {
    MyString(MyString&&, const Alloc&); 
    MyString(const MyString&, const Alloc&); 
};

struct A {
    operator const MyString() const; 
};

int main() {
    MyString s(A{}, Alloc{}); 
    return 0;
}
```

Removing the `const` from `Alloc` for the move case make it work: https://godbolt.org/z/bf3fhPTsc

```cpp
struct Alloc{}; 

struct MyString {
    MyString(MyString&&,  Alloc&); 
    MyString(const MyString&, const Alloc&); 
};

struct A {
    operator const MyString() const; 
};

int main() {
    MyString s(A{}, Alloc{}); 
    return 0;
}
```

Looks somewhat related to: https://github.com/llvm/llvm-project/issues/65137

**Comment by changkhothuychung** [comment]
> Simpler reproducer: https://godbolt.org/z/nqMd4ExPv
> 
> struct Alloc{}; 
> 
> struct MyString {
>     MyString(MyString&&, const Alloc&); 
>     MyString(const MyString&, const Alloc&); 
> };
> 
> struct A {
>     operator const MyString() const; 
> };
> 
> int main() {
>     MyString s(A{}, Alloc{}); 
>     return 0;
> }
> Removing the `const` from `Alloc` for the move case make it work: https://godbolt.org/z/bf3fhPTsc
> 
> struct Alloc{}; 
> 
> struct MyString {
>     MyString(MyString&&,  Alloc&); 
>     MyString(const MyString&, const Alloc&); 
> };
> 
> struct A {
>     operator const MyString() const; 
> };
> 
> int main() {
>     MyString s(A{}, Alloc{}); 
>     return 0;
> }
> Looks somewhat related to: [#65137](https://github.com/llvm/llvm-project/issues/65137)

Oh yeah looks like its the same issue. I have been using lldb to run the program, and I see the move constructor overload is selected, there is an implicit conversion however I think it just checks whether an implicit conversion exists, and later on during initialization step, it fails because `MyString&& r = A{}` is ill-formed. 

Changing `A{}` to `{A{}}` would work because now implicit copy constructor will be applied.

**Comment by changkhothuychung** [comment]
@shafik  I found another case when clang fails but gcc succeeds. 


```cpp
struct A {
};

struct ConvertibleToConstA {
  operator const A();
};

int main()
{
    A a; 
    ConvertibleToConstA ca; 
    a = ca; 
    return 0; 
}
```

Link to compiler explorer - https://godbolt.org/z/79bWMscx9

This code is a test in file clang/test/SemaCXX/copy-assignment.cpp line 85.

**Comment by changkhothuychung** [comment]
I think the main issue here is, `MyString&&` can't bind to the result from `operator const MyString()` due to reference incompability - https://eel.is/c++draft/dcl.init#ref-4
